patches/0006-original_dst_cluster-Avoid-multiple-hosts-for-the-sa.patch

```diff
@@ -0,0 +1,465 @@
++  {
++    absl::WriterMutexLock lock(&host_map_lock_);
++    absl::WriterMutexLock updates_lock(&updates_map_lock_);
++
++    if (updates_map_->empty()) {
++      return; // nothing to do
++    }
++
++    new_hosts.reserve(updates_map_->size());
++
++    for (const auto& [addr, host_use] : *updates_map_) {
++      new_host_map->emplace(addr, host_use);
++      new_hosts.emplace_back(host_use->host_);
++    }
++
++    // Make available for load balancers
++    host_map_ = new_host_map;
++    updates_map_.swap(empty_map);
++  }
++  
+   ASSERT(priority_set_.hostSetsPerPriority().size() == 1);
+   const auto& first_host_set = priority_set_.getOrCreateHostSet(0);
+   HostVectorSharedPtr all_hosts(new HostVector(first_host_set.hosts()));
+-  all_hosts->emplace_back(host);
++  for (auto host : new_hosts) {
++    all_hosts->emplace_back(host);
++  }
+   priority_set_.updateHosts(0,
+                             HostSetImpl::partitionHosts(all_hosts, HostsPerLocalityImpl::empty()),
+-                            {}, {std::move(host)}, {}, absl::nullopt, absl::nullopt);
++                            {}, new_hosts, {}, absl::nullopt, absl::nullopt);
+ }
+ 
+ void OriginalDstCluster::cleanup() {
+-  HostVectorSharedPtr keeping_hosts(new HostVector);
+-  HostVector to_be_removed;
+-  absl::flat_hash_set<absl::string_view> removed_addresses;
+-  auto host_map = getCurrentHostMap();
++  ASSERT_IS_MAIN_OR_TEST_THREAD();
++  const auto* host_map = getHostMap();
++
+   if (!host_map->empty()) {
++    HostVectorSharedPtr keeping_hosts = std::make_shared<HostVector>();
++    HostVector to_be_removed;
++    absl::flat_hash_set<absl::string_view> removed_addresses;
++
+     ENVOY_LOG(trace, "Cleaning up stale original dst hosts.");
+-    for (const auto& [addr, hosts] : *host_map) {
++    for (const auto& [addr, host_use] : *host_map) {
+       // Address is kept in the cluster if either of the two things happen:
+-      // 1) a host has been recently selected for the address; 2) none of the
+-      // hosts are currently in any of the connection pools.
+-      // The set of hosts for a single address are treated as a unit.
++      // 1) a host has been recently selected for the address;
++      // 2) a host is currently in a connection pool.
+       //
+       // Using the used_ bit is preserved for backwards compatibility and to
+       // add a delay between load balancers choosing a host and grabbing a
+@@ -267,49 +347,41 @@ void OriginalDstCluster::cleanup() {
+       // 3) will not delete h since it takes at least one cleanup_interval for
+       // the host to set used_ bit for h to false.
+       bool keep = false;
+-      if (hosts->used_) {
+-        keep = true;
+-        hosts->used_ = false; // Mark to be removed during the next round.
++      if (host_use->used_) {
++	keep = true;
++	host_use->used_ = false; // Mark to be removed during the next round.
+       } else if (Runtime::runtimeFeatureEnabled(
```

Comment:
should we enable this runtime flag in cilium repo?

Reply:
I was confused about this also, but runtime flags defined like this one are `true` by default.

Reply:
https://github.com/envoyproxy/envoy/blob/main/source/common/runtime/runtime_features.cc#L25-L27